### DISCUSSION TITLE
How to override a build time property from extension A in an extension B build step?

### DISCUSSION THREAD

**Discussion by HerrDerb** [quarkusio/quarkus#37487] [Q&A] [answered]
I have an extension B that requires extension A.
Extension A provides some build time config mappings. 
I would like to override certain build time properties from extension A in extension B.

The only way which did work out for me was to define the property in a `application.properties` in extension b.
This is very static. Is it somehow possible to override  a "foreign" build propertie inside of a build step? I could not find anything in the extension guide.

**Comment by mkouba** [comment]
I'm afraid that no such build item exists. And I have to admit that I needed it once too.

CC @radcortez

**Reply by HerrDerb** [reply]
Should I go on then and open a feature request or is there a technical limitation?

**Reply by mkouba** [reply]
That is a question for @radcortez.

**Reply by radcortez** [reply]
`Config` has to be initialized before build steps are executed to provide the method injection points. All injection points are pre-validated before build steps execute, to make sure all the configuration is available. Providing such a build item could render the config invalid mid-flight which I think would be an undesirable effect and potentially even create scenarios where different build steps get different configurations for the same names.

**Reply by HerrDerb** [reply]
Would it be possible to have a buildItem that has (rough idea without any in-depth knowledge) a higher priority as any buildstep that injects a config? Or does the config init happen way before any build item step?
Altough we still would run into a problem if we have two independend extensions which modify the same config 🤔

**Reply by radcortez** [reply]
No, config is initialized and validated before any build step is executed.

> Altough we still would run into a problem if we have two independend extensions which modify the same config

Yes, this is another issue. 

Why do you need to set the config via a build item? Which logic is behind it that couldn't be done in a custom config source?

**Reply by HerrDerb** [reply]
I have a "base" extension which, next to other things, sets a few company wide properties (e.g. special db dialect). My intention was to only set those, if a certain capability (hibernate) is around.
Otherwise, I might get warnings that a unused/unmapped property exists.

**Reply by radcortez** [reply]
One possibility is to look for configuration names that you know that are only available (usually defaults) if the extension is present. You can check for those names and add your own keys in a `ConfigSourceFactory`.

Another possibility is to just query the Classloader for specific classes from those extensions and add your own keys. We follow that approach in the OpenTelemetry Extension.

**Reply by HerrDerb** [reply]
I guess that only works for adding, but not for overriding?

**Reply by radcortez** [reply, marked as answer]
You can override the config if you use the factory.

**Reply by HerrDerb** [reply]
Seems to work, when I use the extension in a project. Altought I wasn't able to write a working deployment test in my extension

**Reply by radcortez** [reply]
> Seems to work, when I use the extension in a project. Altought I wasn't able to write a working deployment test in my extension

Can you please provide more information? What is the issue?

**Reply by HerrDerb** [reply]
I`ll try to make a reproducer in time
